File: unity/Assets/Plugins/Android/UnitySpeech.cs
using UnityEngine;
using System.Collections;

public abstract class UnitySpeech : MonoBehaviour {
    private AndroidJavaClass speech;
    private static bool single = true;
	/// <summary>
    /// Initializes the class (must call from child class's Start() )
    /// </summary>
	public virtual void Start () {
        if(!single) {
            throw new System.Exception("you can only have 1 derived class of UnitySpeech");
            return;
        }
        speech = new AndroidJavaClass("io.github.saltyjeff.unityspeech.UnitySpeech");
        speech.CallStatic("init", gameObject.name);
        single = true;
	}
    /// <summary>
    /// starts listening to user input
    /// </summary>
    public void beginListening () {
        speech.CallStatic("beginListening");
    }
    /// <summary>
    /// will be called when listening begins
    /// </summary>
    public abstract void onSpeechStart();
    private void hasRmsChange(string f) {
        onRmsChange(float.Parse(f));
    }
    /// <summary>
    /// will be called when the RMS changes (still don't know what that is)
    /// </summary>
    /// <param name="value">the new RMS</param>
    public abstract void onRmsChange(float value);
    /// <summary>
    /// will be called when there are partial results
    /// </summary>
    /// <param name="words">array of partials</param>
    public abstract void onPartialResults(string[] words);
    private void hasPartialResults () {
        using (AndroidJavaObject jo = speech.CallStatic<AndroidJavaObject>("getPartials")) {
            onPartialResults(AndroidJNIHelper.ConvertFromJNIArray<string[]>(jo.GetRawObject()));
        }
    }
    /// <summary>
    /// will be called when listening is over and results are in
    /// </summary>
    /// <param name="result">the result</param>
    public abstract void onResult(string result);
    /// <summary>
    /// will be called when the user doesn't have google voice typing installed
    /// </summary>
    public abstract void errNoSpeechRecognition();
    /// <summary>
    /// will be called when the user disabled google voice typing
    /// </summary>
    public abstract void errVoiceTypingDisabled();
    /// <summary>
    /// stops the listening
    /// </summary>
    public void stopListening () {
        speech.CallStatic("stopListening");
    }
    /// <summary>
    /// it's supposed to clear memory leaks, but I don't see much of a problem without it
    /// </summary>
    public void cleanup () {
        speech.CallStatic("cleanup");
    }
    /// <summary>
    /// speech synthesis
    /// </summary>
    /// <param name="words">the words to say</param>
    public void talk(string words) {
        speech.CallStatic("talk", words);
    }
}

File: android/app/src/main/java/io/github/saltyjeff/unityspeech/UnitySpeech.java
package io.github.saltyjeff.unityspeech;

import android.app.Activity;

import com.unity3d.player.UnityPlayer;

import net.gotev.speech.*;

import java.util.List;

public final class UnitySpeech {
    private static String gameObject = "UnitySpeech";
    private static Activity activity;
    private static String[] partials;
    public static final String TAG = "UnitySpeech";
    private UnitySpeech () {
        //its static
    }
    public static void init () {
        activity = UnityPlayer.currentActivity;
        activity.runOnUiThread(new Runnable () {
            @Override
            public void run() {
                Speech.init(activity.getApplication());
                Speech.getInstance().setPreferOffline(true);
            }
        });
    }
    public static void init (String go) {
        gameObject = go;
        init();
    }
    //when there are partials, call this guy to get the array
    public static String[] getPartials() {
        return partials;
    }
    public static void beginListening () {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if(Speech.getInstance().isListening()) {
                    return;
                }
                try {
                    // you must have android.permission.RECORD_AUDIO granted at this point
                    Speech.getInstance().startListening(new SpeechDelegate() {
                        //fires when speech recognition starts
                        @Override
                        public void onStartOfSpeech() {
                            UnityPlayer.UnitySendMessage(gameObject, "onSpeechStart", "");
                        }
                        //fires on whatever RMS is
                        @Override
                        public void onSpeechRmsChanged(float value) {
                            //can someone please expalin what RMS is????
                            UnityPlayer.UnitySendMessage(gameObject, "hasRmsChange", Float.toString(value));
                        }
                        //fires on partial results
                        @Override
                        public void onSpeechPartialResults(List<String> results) {
                            partials = (String[]) results.toArray();
                            UnityPlayer.UnitySendMessage(gameObject, "hasPartialResults", "");
                        }
                        //fires when results are final
                        @Override
                        public void onSpeechResult(String result) {
                            UnityPlayer.UnitySendMessage(gameObject, "onResult", result);
                        }
                    });
                }
                //fires when there's no speech recognition software
                catch (SpeechRecognitionNotAvailable e) {
                    UnityPlayer.UnitySendMessage(gameObject, "errNoSpeechRecognition", "");
                }
                //fires when google voice typing is disabled
                catch (GoogleVoiceTypingDisabledException e) {
                    UnityPlayer.UnitySendMessage(gameObject, "errVoiceTypingDisabled", "");
                }
            }
        });
    }
    //stops listening
    public static void stopListening () {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if(!Speech.getInstance().isListening()) {
                    return;
                }
                Speech.getInstance().stopListening();
            }
        });
    }
    //cleans up
    public static void cleanup () {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Speech.getInstance().unregisterDelegate();
            }
        });
    }
    public static void talk (String words) {
        final String scopeShift = words;
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Speech.getInstance().say(scopeShift);
            }
        });
    }
}
